File: 0147. Insertion Sort List/solution.java
/*
 * My solution with extra space
 *
 * Time: O(n^2)
 * Space: O(n)
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        if ((head == null) || (head.next == null)) return head;
        
        ListNode dummy = new ListNode(Integer.MIN_VALUE);
        ListNode cur = dummy;
        while (head != null) {
            ListNode node = new ListNode(head.val);
            if (head.val >= cur.val) {
                cur.next = node;
                cur = node;
            } else {
                ListNode start = dummy;
                while (start.next.val <= head.val) {
                    start = start.next;
                }
                node.next = start.next;
                start.next = node;
            }
            head = head.next;
        }
        return dummy.next;
    }
}


/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */

File: 0147. Insertion Sort List/solution.py
"""
Other's solution without extra space
(Converted from Java solution)
Time: O(n^2)
Space: O(1)
"""
class Solution:
    def insertionSortList(self, head: ListNode) -> ListNode:
        if head is None or head.next is None:
            return head
        
        dummy = ListNode(head.val);
        dummy.next = head
        prev = dummy
        cur = head
        while (cur is not None):
            if cur.val >= prev.val:
                prev = cur
                cur = cur.next
            else:
                prev.next = cur.next
                start = dummy
                while (start.next.val <= cur.val):
                    start = start.next
                cur.next = start.next
                start.next = cur
                cur = prev.next
        return dummy.next

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
